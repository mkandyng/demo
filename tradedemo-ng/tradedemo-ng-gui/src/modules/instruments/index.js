// dailyTimeSeries/index.js
import * as instrumentActions from "./actions";
import * as instrumentsEpics from "./epics";
import reducer, { NAME } from "./reducer";
import * as components from "./components";

export default { instrumentActions, reducer, NAME, components, instrumentsEpics };
