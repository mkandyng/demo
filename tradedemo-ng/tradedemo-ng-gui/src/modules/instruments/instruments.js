import React from "react";
import {getRandomInt} from "../../libs/utils";

export const MAX_MARKET_FEED_INSTRUMENTS = 5;
export const MAX_INSTRUMENTS = 15;
export const MOUSE_OVER_IMAGE = "img/delete_click.png";
export const MOUSE_OUT_IMAGE = "img/delete.png";

/**
 * [retrieveInstrument return an instrument from the list that matches a given symbol]
 * @param  {[Array]} instruments [Array of instrument objects]
 * @param  {[String]} symbol      [symbol property of the instrument object]
 * @return {[Object]}             [An instrument object from the given array]
 */
export function retrieveInstrument(instruments, symbol) {
  return instruments.find(e => e.symbol === symbol);
}

/**
 * [removeInstrument return a new instruments list and not mutate original array]
 * @param  {[Array]} instruments [Array of instrument objects]
 * @param  {[String]} symbol      [symbol property of the instrument object]
 * @return {[Array]}             [Array of the instrument objects that doesn't contain matched symbol]
 */
export function removeInstrument(instruments, symbol) {
  return instruments.filter(e => e.symbol !== symbol);
}

/**
 * [updateInstrument update an instrument from the gaven instrument if symbol ]
 * @param  {[Array]} instruments      [Array of instrument objects]
 * @param  {[Object]} updateInstrument [A new instrument object that we like to replace]
 * @return {[Array]}                  [Array of the instruments with instrument replaced for matched symbol]
 */
export function updateInstrument(instruments, updateInstrument) {
  return instruments.map(
    e => e.symbol === updateInstrument.symbol
    ? updateInstrument
    : e);
}

export function getDeleteId(instrument) {
  return "deleteImg-" + instrument.symbol;
}

/**
 * [generateMarketfeedMovement Given an instrument, generate a new instrument with different prices]
 * @param  {[Object]} instrument [an instrument object]
 * @return {[type]}            [a new instrument object with new adjust price]
 */
export function generateMarketfeedMovement(instrument) {

  // generate a random margin
  const margin = instrument.price * (0.001 * getRandomInt(1, 10));
  const bidAskMargin = instrument.price * (0.001 * getRandomInt(1, 10));
  const newPrice = calNewPrice(instrument.price, margin);
  const chg = (((newPrice - instrument.open) / instrument.open) * 100).toFixed(2);
  const bid = (newPrice - bidAskMargin).toFixed(2);
  const ask = (newPrice + bidAskMargin).toFixed(2);
  const open = instrument.open * 1.0;
  const last = instrument.price * 1.0;

  return {
    delete: <img id={getDeleteId(instrument)} src={MOUSE_OUT_IMAGE} width={15} height={15} alt={"delete"}/>,
    price: newPrice,
    bidPrice: bid,
    askPrice: ask,
    symbol: instrument.symbol,
    name: instrument.name,
    currency: instrument.currency,
    last: last.toFixed(2),
    open: open.toFixed(2),
    bid: upDownPrice(newPrice, open, bid),
    ask: upDownPrice(newPrice, open, ask),
    chg: upDownPrice(newPrice, open, chg + "%")
  };
}

/**
 * [flashPriceUpdate trigger an update of marketfeed price of one of the instruments]
 * @param  {[Array]} marketfeedInstruments      [Array of instruments in the marketfeed]
 * @param  {[Function]} updateMarketfeedInstrument [Action to change state of marketfeed]
 * @return {[Boolean]}                            [Inform if update successful or not]
 */
export function flashPriceUpdate(marketfeedInstruments, updateMarketfeedInstrument) {
  if (marketfeedInstruments.length > 0) {
    const instrumentIndex = getRandomInt(0, marketfeedInstruments.length - 1);
    const randomInstrument = marketfeedInstruments[instrumentIndex];
    updateMarketfeedInstrument({
      ...randomInstrument,
      bid: "",
      ask: ""
    });
    setTimeout(() => updateMarketfeedInstrument(generateMarketfeedMovement(randomInstrument)), 500);
    return true;
  }
  return false;
};

function calNewPrice(price, margin) {
  return Math.round(Math.random()) === 0
    ? price - margin
    : price + margin;
}

function upDownPrice(price, previousPrice, data) {
  let priceUpDown = "priceUp";
  if (price < previousPrice) {
    priceUpDown = "priceDown";
  }
  return <span className={priceUpDown}>{data}</span>;
}
