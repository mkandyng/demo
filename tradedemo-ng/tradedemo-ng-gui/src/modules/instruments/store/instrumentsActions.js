export const types = {
    FETCH_INSTRUMENTS: "instruments/FETCH_INSTRUMENTS",
    FETCH_INSTRUMENTS_SUCCESS: "instruments/FETCH_INSTRUMENTS_SUCCESS",
    FETCH_INSTRUMENTS_FAILURE: "instruments/FETCH_INSTRUMENTS_FAILURE",
    ADD_INSTRUMENT_TO_MARKETFEED: "instruments/ADD_INSTRUMENT_TO_MARKETFEED",
    ADD_INSTRUMENT_TO_MARKETFEED_SUCCESS: "instruments/ADD_INSTRUMENT_TO_MARKETFEED_SUCCESS",
    ADD_INSTRUMENT_TO_MARKETFEED_FAILURE: "instruments/ADD_INSTRUMENT_TO_MARKETFEED_FAILURE",
    DELETE_MARKETFEED_INSTRUMENT: "instruments/DELETE_MARKETFEED_INSTRUMENT",
    SELECT_MARKETFEED_INSTRUMENT: "instruments/SELECT_MARKETFEED_INSTRUMENT",
    UPDATE_MARKETFEED_INSTRUMENT: "instruments/UPDATE_MARKETFEED_INSTRUMENT"
}

export const fetchInstruments = ()  => ({
    type: types.FETCH_INSTRUMENTS
});

export const fetchInstrumentsSuccess = instruments => ({
    type: types.FETCH_INSTRUMENTS_SUCCESS,
    instruments: instruments
});

export const fetchInstrumentsFailure = error => ({
    type: types.FETCH_INSTRUMENTS_FAILURE,
    error: error
});

export const addInstrumentToMarketfeed = instrument => ({
    type: types.ADD_INSTRUMENT_TO_MARKETFEED,
    instrument: instrument
});

export const addInstrumentToMarketfeedSuccess = instrument => ({
    type: types.ADD_INSTRUMENT_TO_MARKETFEED_SUCCESS,
    instrument: instrument
});

export const addInstrumentToMarketfeedFailure = error => ({
    type: types.ADD_INSTRUMENT_TO_MARKETFEED_FAILURE,
    error: error
});

export const selectMarketfeedInstrument = instrument => ({
    type: types.SELECT_MARKETFEED_INSTRUMENT,
    instrument: instrument
});

export const updateMarketfeedInstrument = instrument => ({
    type: types.UPDATE_MARKETFEED_INSTRUMENT,
    instrument: instrument
});

export const deleteMarketfeedInstrument = instrument => ({
    type: types.DELETE_MARKETFEED_INSTRUMENT,
    instrument: instrument
});
