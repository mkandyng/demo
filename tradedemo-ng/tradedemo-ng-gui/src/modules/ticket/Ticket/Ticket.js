import React from "react";
import { getRandomInt, toggleOpacity, getDateString } from "../../../libs/utils";
import BuySellButton from "../../../components/BuySellButton";
import LabelInput from "../../../components/LabelInput";
import LabelTextArea from "../../../components/LabelTextArea";
import SelectionDropDown from "../../../components/SelectionDropDown";
import submitTicket from "../submitTicket";
import "./ticket.css";

/**
 *
 * [Ticket Component to handle ticket view and interaction]
 * This component is not a pure function because ticket keep track of its own state
 *
 */

class Ticket extends React.Component {

    constructor(props) {
        super(props);
        this.state = { symbol: undefined,
                       orderId:0,
                       quantity: 5,
                       orderType: "Market",
                       price: 0,
                       expiryType: "Day",
                       expiryDate: "",
                       note: "",
                       priceStyle: {opacity: "0.5"},
                       expiryDateStyle: {opacity: "0.5"} };

        this.eventHandler = {
            handleFormChange: e => { const { name, value } = e.target;
                                     this.setState({[name] : value}) },
            orderTypeChange:  (e, instrument) => this.setState({
                      priceStyle: {opacity: toggleOpacity(e.target.value, "Market", true)},
                      orderType: e.target.value}),
            expiryTypeChange: e => this.setState({
                      expiryDateStyle: {opacity: toggleOpacity(e.target.value, "GTD", false)},
                      expiryType: e.target.value,
                      expiryDate: e.target.value === "GTD"?getDateString(new Date(), "dateOnly"):""}),
            handleOnSubmit: (e, ticket, instrument) => {
                      e.preventDefault();
                      const buySell = e.target.innerText;
                      if(submitTicket({ ticket: ticket,
                                        instrument: instrument,
                                        buySell:buySell,
                                        confirmOrder: true,
                                        placeOrder: this.props.placeOrder,
                                        updateOrder: this.props.updateOrder})) {
                          this.setState({orderId: ticket.orderId+1});
                      }
             }
         };
     }

     componentDidMount() {
         const MAX_ORDER_COUNT = 10;
         const interval = setInterval(() => {
             if(this.props.enableDemo && this.state.orderId < MAX_ORDER_COUNT) {
                 let buySell = getRandomInt(0,1) === 0 ? "Buy":"Sell";
                 if(this.props.instrument.symbol && submitTicket({ ticket: this.state,
                                   instrument: this.props.instrument,
                                   buySell: buySell,
                                   confirmOrder: false,
                                   placeOrder: this.props.placeOrder,
                                   updateOrder: this.props.updateOrder })) {

                     this.setState({orderId: this.state.orderId+1});
                 }
             } else {
                clearInterval(interval);
             }
         }, getRandomInt(300,1000));
    }

    componentWillUpdate() {
        if(this.state.symbol !== this.props.instrument.symbol) {
            this.setState({ symbol: this.props.instrument.symbol,
                            price: this.props.instrument.price.toFixed(2) });
        }
    }

    render() {
        return (
            <TicketView ticket={this.state}
                        instrument={this.props.instrument}
                        eventHandler={this.eventHandler} />
        );
    }
}

export function TicketView({ticket, instrument, eventHandler}) {
    return (
        <div id="ticket">
            <form>
                <LabelInput label="Symbol"
                            name="symbol"
                            type="text"
                            value={ticket.symbol}
                            handleOnChange={eventHandler.handleFormChange} />
                <LabelInput label="Quantity"
                            name="quantity"
                            type="number"
                            step={5}
                            value={ticket.quantity}
                            handleOnChange={eventHandler.handleFormChange} />
                <SelectionDropDown id="orderType"
                                   label="Order Type"
                                   name="orderType"
                                   list={["Market","Limit","Stop","Stop Limit"]}
                                   handleOnChange={ e => eventHandler.orderTypeChange(e,instrument) } />
                <LabelInput style={ticket.priceStyle}
                            label="Price"
                            name="price"
                            type="number"
                            step={0.1}
                            value={ticket.price}
                            handleOnChange={eventHandler.handleFormChange} />
                <SelectionDropDown id="expiryType"
                                   label="Expiry Type"
                                   name="expiryType"
                                   list={["Day","GTD","GTC","FOK"]}
                                   handleOnChange={eventHandler.expiryTypeChange} />
                <LabelInput style={ticket.expiryDateStyle}
                            label="Expiry Date"
                            name="expiryDate"
                            type="date"
                            step={0.1}
                            value={ticket.expiryDate}
                            handleOnChange={eventHandler.handleFormChange} />
                <LabelTextArea
                           label="Note"
                           name="note"
                           maxLength={100}
                           value={ticket.note}
                           handleOnChange={eventHandler.handleFormChange} />
                <div>
                    <div id="buySell">
                        <BuySellButton containerId="sellButton"
                                       handleOnClick={e => eventHandler.handleOnSubmit(e, ticket, instrument)}
                                       label={instrument.bidPrice}
                                       id="sell"
                                       buttonName="Sell" />
                        <BuySellButton containerId="buyButton"
                                       handleOnClick={e => eventHandler.handleOnSubmit(e, ticket, instrument)}
                                       label={instrument.askPrice}
                                       id="buy"
                                       buttonName="Buy" />
                    </div>
                </div>
            </form>
        </div>
    )
}

export default Ticket;
