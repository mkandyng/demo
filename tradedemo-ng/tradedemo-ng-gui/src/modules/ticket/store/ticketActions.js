export const types = {
    UPDATE_TICKET: "ticket/UPDATE"
};

export const updateTicket = ticketProps => ({
    type: types.UPDATE_TICKET,
    ticketProps: ticketProps
});
