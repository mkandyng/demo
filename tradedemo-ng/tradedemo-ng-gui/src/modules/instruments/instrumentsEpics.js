import "rxjs/add/operator/switchMap";
import "rxjs/add/operator/concatMap";
import "rxjs/add/operator/map";
import "rxjs/add/observable/of";
import "rxjs/add/operator/catch";
import { ajax } from "rxjs/observable/dom/ajax";
import { instrumentServiceUrl } from "../../libs/resources";
import { MAX_MARKET_FEED_INSTRUMENTS } from "./index"
import * as instrumentsActions from "./instrumentsActions";
import timeSeries from "../timeSeries";
import generateMarketfeedMovement from "./generateMarketfeedMovement";

export const fetchInstrumentsEpic = function(action$) {
    const fetchInstruments = () => {
      return ajax
          .getJSON(instrumentServiceUrl + "/instruments")
          .map(payload => payload.map((instrument) => {
              let object = {
                  symbol: instrument["symbol"],
                  name: instrument["name"],
                  currency: instrument["currency"]
              }
              return object;
          }));
    }

    return action$
        .ofType(instrumentsActions.types.FETCH_INSTRUMENTS)
        .switchMap(action =>
            fetchInstruments()
            .concatMap(instruments => [
              instrumentsActions.fetchInstrumentsSuccess(instruments),
              ...instruments.slice(0,MAX_MARKET_FEED_INSTRUMENTS)
                    .map(instrument => instrumentsActions.addInstrumentToMarketfeed(instrument))
            ])
            .catch(error => {instrumentsActions.fetchInstrumentsFailure(error.message)})
        )
}

export const addInstrumentToMarketfeedEpic = function(action$) {
    const fetchInstrumentQuote = action => {
      return ajax
          .getJSON(instrumentServiceUrl + "/instrumentQuote/" + action.instrument.symbol)
          .map(data => {
                  const newInstrument = { symbol:   action.instrument.symbol,
                                          name:     action.instrument.name,
                                          currency: action.instrument.currency,
                                          price:    data["price"],
                                          open:     data["open"] };
                  return (
                     generateMarketfeedMovement(newInstrument)
                  )
          });
    }

    return action$
        .ofType(instrumentsActions.types.ADD_INSTRUMENT_TO_MARKETFEED)
        .concatMap(action => fetchInstrumentQuote(action)
          .concatMap(instrument => [
              instrumentsActions.addInstrumentToMarketfeedSuccess(instrument),
              instrumentsActions.selectMarketfeedInstrument(instrument)
            ]
          )
          .catch(error => {instrumentsActions.addInstrumentToMarketfeedFailure(error)})
    )
}

export const selectMarketfeedInstrumentEpic = function(action$) {
    return action$
        .ofType(instrumentsActions.types.SELECT_MARKETFEED_INSTRUMENT)
        .concatMap(action => [
              timeSeries.timeSeriesActions.fetchIntradayTimeSeries(action.instrument.symbol),
              timeSeries.timeSeriesActions.fetchDailyTimeSeries(action.instrument.symbol)
            ]
        )
}
