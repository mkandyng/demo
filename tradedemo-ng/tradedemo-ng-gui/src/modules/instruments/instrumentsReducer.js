import * as instrumentActions from "./instrumentsActions";

export const NAME = "instruments";

const removeInstrument = (instruments, symbol) => instruments.filter(e => e.symbol !== symbol);
const updateInstrument = (instruments, updateInstrument) =>
                          instruments.map(e => e.symbol === updateInstrument.symbol ? updateInstrument: e);

export default function instrumentsReducer(state = { instruments:[],
                                                     marketfeedInstruments:[],
                                                     selected:{} }, action) {

    switch (action.type) {
        case instrumentActions.types.FETCH_INSTRUMENTS_SUCCESS:
            return { instruments: action.payload,
                     marketfeedInstruments:[],
                     selected:{} }
        case instrumentActions.types.FETCH_INSTRUMENTS_FAILURE:
            return state;
        case instrumentActions.types.ADD_INSTRUMENT_TO_MARKETFEED_SUCCESS:
            return { instruments: removeInstrument(state.instruments, action.instrument.symbol),
                     marketfeedInstruments:[action.instrument, ...state.marketfeedInstruments],
                     selected: action.instrument };
        case instrumentActions.types.ADD_INSTRUMENT_TO_MARKETFEED_FAILURE:
            return state;
        case instrumentActions.types.SELECT_MARKETFEED_INSTRUMENT:
            return { ...state,
                     selected: action.instrument };
        case instrumentActions.types.UPDATE_MARKETFEED_INSTRUMENT:
            const selected = state.selected.symbol === action.instrument.symbol?action.instrument:state.selected;
            return { ...state,
                     marketfeedInstruments: updateInstrument(state.marketfeedInstruments, action.instrument),
                     selected: selected };
        case instrumentActions.types.DELETE_MARKETFEED_INSTRUMENT:
            const  marketfeedInstruments = removeInstrument(state.marketfeedInstruments, action.instrument.symbol);
            return { instruments: [action.instrument, ...state.instruments],
                     marketfeedInstruments:marketfeedInstruments,
                     selected: marketfeedInstruments.length === 0 ? {} : marketfeedInstruments[0] };
        default:
           return state;
    }
}
