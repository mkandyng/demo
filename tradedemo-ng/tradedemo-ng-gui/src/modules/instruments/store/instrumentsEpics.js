import "rxjs/add/operator/switchMap";
import "rxjs/add/operator/concatMap";
import "rxjs/add/operator/map";
import "rxjs/add/observable/of";
import "rxjs/add/operator/catch";
import { combineEpics } from "redux-observable";
import { Observable } from 'rxjs';
import { catchError, concatMap } from 'rxjs/operators'
import { instrumentServiceUrl } from "../../../libs/resources";
import { fetchIntradayTimeSeries,
         fetchDailyTimeSeries } from "../../timeSeries/store/timeSeriesActions";
import * as instrumentsActions from "./instrumentsActions";
import { generateMarketfeedMovement, MAX_MARKET_FEED_INSTRUMENTS } from "../instruments";

export const instrumentsEpics = combineEpics( fetchInstrumentsEpic,
                                              addInstrumentToMarketfeedEpic,
                                              selectMarketfeedInstrumentEpic )

function fetchInstrumentsEpic(action$, store, {ajax}) {
    function fetchInstruments(ajaxRestApi){
      return ajaxRestApi
          .getJSON(instrumentServiceUrl + "/instruments/")
          .map(payload => payload.map(instrument => {
              let object = {
                  symbol: instrument.symbol,
                  name: instrument.name,
                  currency: instrument.currency
              }
              return object;
          }));
    }

    return action$
        .ofType(instrumentsActions.types.FETCH_INSTRUMENTS)
        .switchMap(action =>
            fetchInstruments(ajax).pipe(
                concatMap(instruments => [
                            instrumentsActions.fetchInstrumentsSuccess(instruments),
                            ...instruments.slice(0,MAX_MARKET_FEED_INSTRUMENTS)
                                  .map(instrument => instrumentsActions.addInstrumentToMarketfeed(instrument))
                ]),
                catchError(error => Observable.of(instrumentsActions.fetchInstrumentsFailure(error)))
          )
        )
}

function addInstrumentToMarketfeedEpic(action$, store, {ajax}) {
    const fetchInstrumentQuote = action => {
      return ajax
          .getJSON(instrumentServiceUrl + "/instrumentQuote/" + action.instrument.symbol)
          .map(data => {
                  const newInstrument = { symbol:   action.instrument.symbol,
                                          name:     action.instrument.name,
                                          currency: action.instrument.currency,
                                          price:    data.price,
                                          open:     data.open };
                  return (
                     generateMarketfeedMovement(newInstrument)
                  )
          });
    }

    return action$
        .ofType(instrumentsActions.types.ADD_INSTRUMENT_TO_MARKETFEED)
        .concatMap(action => fetchInstrumentQuote(action).pipe(
            concatMap(instrument => [
                instrumentsActions.addInstrumentToMarketfeedSuccess(instrument),
                instrumentsActions.selectMarketfeedInstrument(instrument) ]
            ), catchError(error => Observable.of(instrumentsActions.addInstrumentToMarketfeedFailure(error)))
          )
    )
}

function selectMarketfeedInstrumentEpic(action$) {
    return action$
        .ofType(instrumentsActions.types.SELECT_MARKETFEED_INSTRUMENT)
        .concatMap(action => [
              fetchIntradayTimeSeries(action.instrument.symbol),
              fetchDailyTimeSeries(action.instrument.symbol)
            ]
        )
}
