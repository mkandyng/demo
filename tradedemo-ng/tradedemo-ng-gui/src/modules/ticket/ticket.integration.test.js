import React from "react";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { combineReducers } from "redux";
import { getDateString,
         padDigits,
         createStoreWithMiddleware } from "../../libs/utils";
import { MAX_INSTRUMENTS } from "../instruments/instruments";
import * as orderbook from "../../modules/orderbook/store/orderbookReducer";
import * as instruments from "../../modules/instruments/store/instrumentsReducer";
import { instrumentsEpics } from "../../modules/instruments/store/instrumentsEpics";
import { placeOrder, updateOrder } from "../../modules/orderbook/store/orderbookActions"
import { fetchInstruments } from "../../modules/instruments/store/instrumentsActions";
import { createInstruments,
         getJSONFunction } from "../instruments/instruments.test.helpers";
import Ticket from "./Ticket";

/**
 *
 * Test of Ticket component integrate with Redux store to verify against orderbook
 *
 */

describe("ticket integration tests", () => {
    let store = undefined;
    let component = undefined;
    let props = undefined;

    beforeEach(() => {
        store = createStoreWithMiddleware({getJSON: url => getJSONFunction(url, createInstruments(MAX_INSTRUMENTS))},
                                           combineReducers({
                                              [orderbook.NAME]:orderbook.orderbookReducer,
                                              [instruments.NAME]: instruments.instrumentsReducer
                                            }),
                                            instrumentsEpics);
        store.dispatch(fetchInstruments());
        props = {
            instrument: store.getState().instruments.selected,
            placeOrder: order => store.dispatch(placeOrder(order)),
            updateOrder: order => store.dispatch(updateOrder(order)),
            enableDemo: false
        }
        component = mount(<Ticket {... props} />);
    });

    it("should matchSnapshot for mount Ticket component ", () => {
        expect(toJson(component)).toMatchSnapshot();
    });

    it("should handle standard Input change ", () => {
        verifyTicketChange("input", "quantity", 99);
        verifyTicketChange("input", "price", "1.02");
        verifyTicketChange("textarea", "note", "note value");
        verifyTicketChange("input", "expiryDate", "2019-10-11");
    });

    it("should handle orderType change", () => {
        const verifyOrderType = (orderType, expectedOpacity, instrument) => {
            verifyTicketChange("select", "orderType", orderType);
            expect(component.state().priceStyle).toStrictEqual({"opacity": expectedOpacity});
        }
        const instrument = store.getState().instruments.marketfeedInstruments[0];
        verifyOrderType("Market", 0.5, instrument);
        verifyOrderType("Limit", 1.0, instrument);
        verifyOrderType("Stop", 1.0, instrument);
        verifyOrderType("Stop Limit", 1.0, instrument);
    });

    it("should handle expiryType change", () => {
        const verifyExpiryType = (expiryType, expectedOpacity, expectedExpiryDate) => {
            verifyTicketChange("select", "expiryType", expiryType);
            expect(component.state().expiryDateStyle).toStrictEqual({"opacity": expectedOpacity});
            expect(component.state().expiryDate).toStrictEqual(expectedExpiryDate);
        }
        verifyExpiryType("GTD", 1.0, getDateString(new Date(), "dateOnly"));
        verifyExpiryType("Day", 0.5, "");
        verifyExpiryType("FOK", 0.5, "");
        verifyExpiryType("GTC", 0.5, "");
    });

    it("should handle buy/sell submit click to place order", () => {
        window.confirm = jest.fn(() => true);
        jest.useFakeTimers();
        jest.runAllTimers();
        const verifyBuySellButtonClickSubmitOrder = (buySellId, buySell) => {
            // Given
            const expectedOrderRef = "XA"+ padDigits(component.state().orderId, 8);
            const ticket = mount(<Ticket {... props} ticket={component.state()}/>)

            // when
            ticket.find("div#" + buySellId)
                  .simulate("click", { target: { innerText: buySell }});

            // Then
            const newOrder = store.getState().orderbook.find(
                                  order=>order.orderRef === expectedOrderRef);
            expect(newOrder).not.toBeUndefined();
            expect(newOrder.buySell).toStrictEqual(buySell);
        }
        verifyBuySellButtonClickSubmitOrder("buyButton", "Buy");
        verifyBuySellButtonClickSubmitOrder("sellButton", "Sell");
    });

    it("should handle buy/sell submit click when confirm is false, no order placed", () => {
        window.confirm = jest.fn(() => false);
        verifyNoOrderPlaced("buyButton", "Buy");
        verifyNoOrderPlaced("sellButton", "Sell");
    });

    it("should handle buy/sell submit click when quantity zero is rejected, no order placed", () => {
        window.confirm = jest.fn(() => true);
        verifyNoOrderPlaced("buyButton", "Buy", {quantity: 0});
        verifyNoOrderPlaced("sellButton", "Sell", {quantity: 0});
        expect(window.alert).toHaveBeenCalled();
    });

    const verifyNoOrderPlaced = (buySellId, buySell, ticketPropsChange = {}) => {
        // Given
        const ticket = mount(<Ticket {... props}/>)
        ticket.setState(ticketPropsChange);

        // when
        ticket.find("div#" + buySellId)
              .simulate("click", { target: { innerText: buySell }});

        // Then
        expect(store.getState().orderbook.length).toStrictEqual(0);
    }

    const verifyTicketChange = (elementTag, inputName, changeValue) => {
        // Given
        const element = component.find(elementTag).find({ name: inputName })

        // when
        element.simulate('change', { target: { name: inputName, value: changeValue }});

        // Then
        expect(component.state()[inputName]).toStrictEqual(changeValue);
    }
});
