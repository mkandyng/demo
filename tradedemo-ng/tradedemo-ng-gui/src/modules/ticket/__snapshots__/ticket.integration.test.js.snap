// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ticket integration tests should matchSnapshot for mount Ticket component  1`] = `
<Ticket
  enableDemo={false}
  instrument={
    Object {
      "currency": "USD",
      "name": "instrument4",
      "open": 1,
      "price": 1,
      "symbol": "symbol4",
    }
  }
  placeOrder={[Function]}
  updateOrder={[Function]}
>
  <TicketView
    eventHandler={
      Object {
        "expiryTypeChange": [Function],
        "handleFormChange": [Function],
        "handleOnSubmit": [Function],
        "orderTypeChange": [Function],
      }
    }
    instrument={
      Object {
        "currency": "USD",
        "name": "instrument4",
        "open": 1,
        "price": 1,
        "symbol": "symbol4",
      }
    }
    ticket={
      Object {
        "expiryDate": "",
        "expiryDateStyle": Object {
          "opacity": "0.5",
        },
        "expiryType": "Day",
        "note": "",
        "orderId": 0,
        "orderType": "Market",
        "price": 0,
        "priceStyle": Object {
          "opacity": "0.5",
        },
        "quantity": 5,
        "symbol": undefined,
      }
    }
  >
    <div
      id="ticket"
    >
      <form>
        <LabelInput
          handleOnChange={[Function]}
          label="Symbol"
          name="symbol"
          type="text"
        >
          <p>
            <label>
              Symbol
              :
              <input
                name="symbol"
                onChange={[Function]}
                type="text"
                value=""
              />
            </label>
          </p>
        </LabelInput>
        <LabelInput
          handleOnChange={[Function]}
          label="Quantity"
          name="quantity"
          step={5}
          type="number"
          value={5}
        >
          <p>
            <label>
              Quantity
              :
              <input
                name="quantity"
                onChange={[Function]}
                step={5}
                type="number"
                value={5}
              />
            </label>
          </p>
        </LabelInput>
        <SelectionDropDown
          handleOnChange={[Function]}
          id="orderType"
          label="Order Type"
          list={
            Array [
              "Market",
              "Limit",
              "Stop",
              "Stop Limit",
            ]
          }
          name="orderType"
        >
          <p>
            <label>
              Order Type
              :
              <select
                id="orderType"
                name="orderType"
                onChange={[Function]}
              >
                <option
                  id="orderTypeMarket"
                  key="orderTypeMarket"
                  value="Market"
                >
                  Market
                </option>
                <option
                  id="orderTypeLimit"
                  key="orderTypeLimit"
                  value="Limit"
                >
                  Limit
                </option>
                <option
                  id="orderTypeStop"
                  key="orderTypeStop"
                  value="Stop"
                >
                  Stop
                </option>
                <option
                  id="orderTypeStop Limit"
                  key="orderTypeStop Limit"
                  value="Stop Limit"
                >
                  Stop Limit
                </option>
              </select>
            </label>
          </p>
        </SelectionDropDown>
        <LabelInput
          handleOnChange={[Function]}
          label="Price"
          name="price"
          step={0.1}
          style={
            Object {
              "opacity": "0.5",
            }
          }
          type="number"
          value={0}
        >
          <p
            style={
              Object {
                "opacity": "0.5",
              }
            }
          >
            <label>
              Price
              :
              <input
                name="price"
                onChange={[Function]}
                step={0.1}
                type="number"
                value=""
              />
            </label>
          </p>
        </LabelInput>
        <SelectionDropDown
          handleOnChange={[Function]}
          id="expiryType"
          label="Expiry Type"
          list={
            Array [
              "Day",
              "GTD",
              "GTC",
              "FOK",
            ]
          }
          name="expiryType"
        >
          <p>
            <label>
              Expiry Type
              :
              <select
                id="expiryType"
                name="expiryType"
                onChange={[Function]}
              >
                <option
                  id="expiryTypeDay"
                  key="expiryTypeDay"
                  value="Day"
                >
                  Day
                </option>
                <option
                  id="expiryTypeGTD"
                  key="expiryTypeGTD"
                  value="GTD"
                >
                  GTD
                </option>
                <option
                  id="expiryTypeGTC"
                  key="expiryTypeGTC"
                  value="GTC"
                >
                  GTC
                </option>
                <option
                  id="expiryTypeFOK"
                  key="expiryTypeFOK"
                  value="FOK"
                >
                  FOK
                </option>
              </select>
            </label>
          </p>
        </SelectionDropDown>
        <LabelInput
          handleOnChange={[Function]}
          label="Expiry Date"
          name="expiryDate"
          step={0.1}
          style={
            Object {
              "opacity": "0.5",
            }
          }
          type="date"
          value=""
        >
          <p
            style={
              Object {
                "opacity": "0.5",
              }
            }
          >
            <label>
              Expiry Date
              :
              <input
                name="expiryDate"
                onChange={[Function]}
                step={0.1}
                type="date"
                value=""
              />
            </label>
          </p>
        </LabelInput>
        <LabelTextArea
          handleOnChange={[Function]}
          label="Note"
          maxLength={100}
          name="note"
          value=""
        >
          <p>
            <label>
              Note
              <textarea
                maxLength={100}
                name="note"
                onChange={[Function]}
                value=""
              />
            </label>
          </p>
        </LabelTextArea>
        <div>
          <div
            id="buySell"
          >
            <BuySellButton
              buttonName="Sell"
              containerId="sellButton"
              handleOnClick={[Function]}
              id="sell"
            >
              <div
                id="sellButton"
                onClick={[Function]}
              >
                <label />
                <button
                  id="sell"
                >
                  Sell
                </button>
              </div>
            </BuySellButton>
            <BuySellButton
              buttonName="Buy"
              containerId="buyButton"
              handleOnClick={[Function]}
              id="buy"
            >
              <div
                id="buyButton"
                onClick={[Function]}
              >
                <label />
                <button
                  id="buy"
                >
                  Buy
                </button>
              </div>
            </BuySellButton>
          </div>
        </div>
      </form>
    </div>
  </TicketView>
</Ticket>
`;
